plugins/feeds/public/ViruSign.py

```diff
@@ -0,0 +1,64 @@
+import csv
+import logging
+import requests
+from datetime import datetime, timedelta
+from core.observables import Hash
+from core.feed import Feed
+from core.config.config import yeti_config
+from core.errors import ObservableValidationError
+
+class ViruSign(Feed):
+
+    default_values = {
+        "frequency": timedelta(hours=24),
+        "name": "ViruSign",
+        "source": "https://www.virusign.com/get_hashlist.php?ssdeep&imphash&sha256&sha1&md5&n=ANY&start_date={date}&end_date={date}",
+        "description": "This feed contains daily list of hashes from ViruSign",
+    }
+
+    def update(self):
+        today = datetime.now().strftime("%Y-%m-%d")
+        try:
+            r = requests.get(self.source.format(date=today), headers={"User-Agent": "yeti-project"})
+            if r.ok:
+                reader = csv.reader(r.content.splitlines(), quotechar='"')
+                for line in reader:
+                    self.analyze(line)
+        except Exception as e:
+            logging.error(e)
+
+    def analyze(self, line):
+        ssdeep, imphash, sha256, sha1, md5 = line
+        context = {}
+        context['date_added'] = datetime.now()
+        context['source'] = self.name
+
+        try:
+            md5_data = Hash.get_or_create(value=md5)
+            if md5_data.new is True or self.name not in md5_data.sources:
+                md5_data.add_context(context)
+                md5_data.add_source(self.name)
+        except ObservableValidationError as e:
+            logging.error(e)
+
+        try:
+            sha1_data = Hash.get_or_create(value=sha1)
+            if sha1_data.new is True or self.name not in sha1_data.sources:
+                sha1_data.add_context(context)
+                sha1_data.add_source(self.name)
+        except ObservableValidationError as e:
+            logging.error(e)
+
+        try:
+            sha256_data = Hash.get_or_create(value=sha256)
+            if sha256_data.new is True or self.name not in sha256_data.sources:
+                sha256_data.add_context(context)
+                sha256_data.add_source(self.name)
+        except ObservableValidationError as e:
+            logging.error(e)
+
+        try:
+            md5_data.active_link_to(sha1_data, 'sha1', self.name)
```

Comment:
I see what you're doing here. I think the best way would be to actually create a File observbale, and link all hashes to that file.

Reply:
any tip how to do that quickly?

Reply:
Have a look at `hybrid_analysis.py`:

```python

from core.observables import Hash, Hostname, File

# [...]

    def analyze(self, item):
        f_hyb = File.get_or_create(value='FILE:{}'.format(item['sha256']))

        sha256 = Hash.get_or_create(value=item['sha256'])
        f_hyb.active_link_to(sha256, 'sha256', self.name)
```

So basically you create a `File` object, and you link all hashes to that file object instead of linking them to each other.
The `File` object is pretty abstract, so its value is just a unique way of describing the object: we use `file:`  + its sha56 hash.